File: scripts/inspect_pmp.py
import sqlite3
import os

file_path = r"D:\Code Antinigaty\Phan mem quan ly file V4\Du_an_165.pmp"

if not os.path.exists(file_path):
    print(f"File not found: {file_path}")
    exit(1)

# Check first 16 bytes for SQLite signature
with open(file_path, "rb") as f:
    header = f.read(16)
    print(f"Header: {header.hex(' ')}")
    if header.startswith(b"SQLite format 3\x00"):
        print("Format: SQLite")
    else:
        print("Format: Unknown/Binary")

# If SQLite, list tables and schema
try:
    conn = sqlite3.connect(file_path)
    cursor = conn.cursor()
    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
    tables = cursor.fetchall()
    print(f"Tables: {[t[0] for t in tables]}")
    for table_name in [t[0] for t in tables]:
        print(f"\n--- Schema for table: {table_name} ---")
        cursor.execute(f"PRAGMA table_info({table_name});")
        columns = cursor.fetchall()
        for col in columns:
            print(col)
        
        # Đặc biệt kiểm tra bảng metadata nếu có
        if table_name == 'metadata':
             cursor.execute("SELECT key, length(value) FROM metadata;")
             metadata_keys = cursor.fetchall()
             print(f"Metadata keys: {metadata_keys}")

        # Sample data
        try:
            cursor.execute(f"SELECT * FROM {table_name} LIMIT 3;")
            rows = cursor.fetchall()
            print(f"Sample data ({table_name}): {rows}")
        except Exception as e:
            print(f"Could not read data from {table_name}: {e}")
            
    conn.close()
except Exception as e:
    print(f"Error inspecting SQLite: {e}")

File: BAK/src-tauri/src/bin/inspect_pmp.rs
use rusqlite::Connection;

fn main() {
    if let Err(e) = run() {
        eprintln!("ERROR: {:?}", e);
        std::process::exit(1);
    }
}

fn run() -> Result<(), Box<dyn std::error::Error>> {
    let path = "D:\\Code Antinigaty\\Phan mem quan ly file V4\\Du_an_165.pmp";
    println!("Inspecting file: {}", path);
    let conn = Connection::open(path)?;

    let mut stmt = conn.prepare("SELECT name FROM sqlite_master WHERE type='table'")?;
    let tables: Vec<String> = stmt
        .query_map([], |row| row.get(0))?
        .filter_map(|r| r.ok())
        .collect();

    println!("Tables found: {:?}", tables);

    for table in &[
        "v1_design_events",
        "design_events",
        "event_store",
        "v1_projects",
        "projects",
    ] {
        if tables.contains(&table.to_string()) {
            println!("\n--- Table: {} ---", table);
            let count: i64 = conn
                .query_row(&format!("SELECT COUNT(*) FROM {}", table), [], |r| r.get(0))
                .unwrap_or(0);
            println!("Count: {}", count);

            if count > 0 {
                let mut stmt = conn.prepare(&format!("SELECT * FROM {} LIMIT 3", table))?;
                let col_count = stmt.column_count();
                println!("Columns: {:?}", stmt.column_names());

                println!("Sample (top 3):");
                let mut rows = stmt.query([])?;
                while let Some(row) = rows.next()? {
                    let mut vals = Vec::new();
                    for i in 0..col_count {
                        let val = match row.get::<_, rusqlite::types::Value>(i)? {
                            rusqlite::types::Value::Null => "NULL".to_string(),
                            rusqlite::types::Value::Integer(i) => i.to_string(),
                            rusqlite::types::Value::Real(f) => f.to_string(),
                            rusqlite::types::Value::Text(s) => s,
                            rusqlite::types::Value::Blob(b) => format!("<blob {} bytes>", b.len()),
                        };
                        vals.push(val);
                    }
                    println!("  {}", vals.join(" | "));
                }

                if table.contains("events") || table.contains("store") {
                    println!("Project IDs in this table:");
                    if let Ok(mut stmt) =
                        conn.prepare(&format!("SELECT DISTINCT project_id FROM {}", table))
                    {
                        let mut rows = stmt.query([])?;
                        while let Some(row) = rows.next()? {
                            let id: rusqlite::types::Value = row.get(0)?;
                            println!("  - {:?}", id);
                        }
                    } else if let Ok(mut stmt) =
                        conn.prepare(&format!("SELECT DISTINCT project_uuid FROM {}", table))
                    {
                        let mut rows = stmt.query([])?;
                        while let Some(row) = rows.next()? {
                            let id: rusqlite::types::Value = row.get(0)?;
                            println!("  - {:?}", id);
                        }
                    }
                }
            }
        }
    }

    Ok(())
}
